IRGen: always execute the coroutine lowering pass

compnerd opened the pull request:
Do not gate the coroutine extension points on the LLVM passes.  Without
running this pass, the generated IR cannot be used by the LLVM tooling.
This allows generating the LLVM IR to debug issues in the LLVM backend.

I encountered this when trying to isolate a debug info generation bug
which seems to be caused by the SRoA pass in LLVM.  By allowing to emit
the LLVM IR without the LLVM optimizations, it is possible to isolate
the LLVM pass operation via `opt` from LLVM.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves SR-NNNN.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
CC: @aschwaighofer @rjmccall

compnerd commented:
@swift-ci please test

compnerd commented:
I suppose that for the tests we still need another way to prevent the lowering of the coroutines :/

aschwaighofer commented:
As you noted. It is very useful to turn on/off the coroutine passes for debugging. Maybe we introduce a DisableMandatorySwiftSpecificOptimizations flag.

compnerd commented:
@aschwaighofer - sure, that sounds good to me, especially since its primary intention is for testing.

compnerd committed:
IRGen: always execute the coroutine lowering pass

Do not gate the coroutine extension points on the LLVM passes. Without
running this pass, the generated IR cannot be used by the LLVM tooling.
This allows generating the LLVM IR to debug issues in the LLVM backend.

I encountered this when trying to isolate a debug info generation bug
which seems to be caused by the SRoA pass in LLVM. By allowing to emit
the LLVM IR without the LLVM optimizations, it is possible to isolate
the LLVM pass operation via `opt` from LLVM.

compnerd commented:
@aschwaighofer - rather than introduce a new flag, I decided to base it on the existing `-disable-swift-specific-llvm-optzns`.  This `-disable-swift-specific-llvm-optzns` effectively spells DisableMandatorySwiftSpecificOptimizations (previously it aliased `-disable-llvm-optzns`).

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test macOS platform

shahmishal commented:
@swift-ci please clean test macOS platform

compnerd commented:
@swift-ci please test windows platform
